File: artifacts/api-server/src/routes/students.ts
import { Router, type IRouter } from "express";
import { eq, ilike, and, sql } from "drizzle-orm";
import { db, studentsTable, enrollmentsTable, coursesTable, assignmentsTable, gradesTable } from "@workspace/db";
import type { GradingScheme } from "@workspace/db";
import {
  ListStudentsQueryParams,
  CreateStudentBody,
  GetStudentParams,
  UpdateStudentParams,
  UpdateStudentBody,
  DeleteStudentParams,
} from "@workspace/api-zod";
import { scoreToLetter, toPercent, letterToPoints } from "../lib/gradeUtils";
import { StudentModel } from "../domain/StudentModel";
import type { AssignmentScore } from "../domain/GradeBook";

const router: IRouter = Router();

// Helper: compute GPA for a student using per-course GradeBook
async function computeStudentGpa(studentId: number): Promise<number | null> {
  const enrollments = await db
    .select({ courseId: enrollmentsTable.courseId })
    .from(enrollmentsTable)
    .where(eq(enrollmentsTable.studentId, studentId));

  if (enrollments.length === 0) return null;

  const courseGrades = [];

  for (const enr of enrollments) {
    const course = await db
      .select()
      .from(coursesTable)
      .where(eq(coursesTable.id, enr.courseId))
      .limit(1);
    if (!course[0]) continue;

    const assignments = await db
      .select()
      .from(assignmentsTable)
      .where(eq(assignmentsTable.courseId, enr.courseId));

    const scores: AssignmentScore[] = [];
    for (const asgn of assignments) {
      const gradeRows = await db
        .select()
        .from(gradesTable)
        .where(
          and(
            eq(gradesTable.studentId, studentId),
            eq(gradesTable.assignmentId, asgn.id),
          ),
        )
        .limit(1);
      if (gradeRows[0]) {
        scores.push({
          assignmentId: asgn.id,
          score: Number(gradeRows[0].score),
          maxScore: Number(asgn.maxScore),
          weight: Number(asgn.weight),
          name: asgn.name,
          type: asgn.type,
        });
      }
    }

    const gradeInfo = StudentModel.computeCourseGrade(
      scores,
      course[0].id,
      course[0].name,
      course[0].code,
      course[0].credits,
      course[0].semester,
      course[0].gradingScheme as GradingScheme,
    );
    courseGrades.push(gradeInfo);
  }

  // Use StudentModel to calculate GPA (excludes pass/fail courses)
  const dummy = new StudentModel(studentId, "", "", "", 0, "");
  return dummy.calculateGPA(courseGrades);
}

router.get("/students", async (req, res): Promise<void> => {
  const query = ListStudentsQueryParams.safeParse(req.query);
  if (!query.success) {
    res.status(400).json({ error: query.error.message });
    return;
  }

  const { search, major, year } = query.data;

  const conditions = [];
  if (search) conditions.push(ilike(studentsTable.name, `%${search}%`));
  if (major) conditions.push(ilike(studentsTable.major, `%${major}%`));
  if (year !== undefined) conditions.push(eq(studentsTable.year, year));

  const students = await db
    .select()
    .from(studentsTable)
    .where(conditions.length > 0 ? and(...conditions) : undefined)
    .orderBy(studentsTable.name);

  const result = await Promise.all(
    students.map(async (s) => ({
      ...s,
      gpa: await computeStudentGpa(s.id),
      created_at: s.createdAt.toISOString(),
      student_id: s.studentId,
    })),
  );

  res.json(result);
});

router.post("/students", async (req, res): Promise<void> => {
  const parsed = CreateStudentBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const { name, email, student_id, year, major } = parsed.data as { name: string; email: string; student_id: string; year: number; major: string };
  const [student] = await db
    .insert(studentsTable)
    .values({ name, email, studentId: student_id, year, major })
    .returning();

  res.status(201).json({ ...student, student_id: student.studentId, gpa: null, created_at: student.createdAt.toISOString() });
});

router.get("/students/:id", async (req, res): Promise<void> => {
  const params = GetStudentParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }

  const [student] = await db
    .select()
    .from(studentsTable)
    .where(eq(studentsTable.id, params.data.id));

  if (!student) {
    res.status(404).json({ error: "Student not found" });
    return;
  }

  const gpa = await computeStudentGpa(student.id);
  res.json({ ...student, student_id: student.studentId, gpa, created_at: student.createdAt.toISOString() });
});

router.put("/students/:id", async (req, res): Promise<void> => {
  const params = UpdateStudentParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }

  const parsed = UpdateStudentBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const updates: Record<string, unknown> = {};
  const d = parsed.data as Record<string, unknown>;
  if (d.name) updates.name = d.name;
  if (d.email) updates.email = d.email;
  if (d.year !== undefined) updates.year = d.year;
  if (d.major) updates.major = d.major;

  const [student] = await db
    .update(studentsTable)
    .set(updates)
    .where(eq(studentsTable.id, params.data.id))
    .returning();

  if (!student) {
    res.status(404).json({ error: "Student not found" });
    return;
  }

  const gpa = await computeStudentGpa(student.id);
  res.json({ ...student, student_id: student.studentId, gpa, created_at: student.createdAt.toISOString() });
});

router.delete("/students/:id", async (req, res): Promise<void> => {
  const params = DeleteStudentParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }

  await db.delete(studentsTable).where(eq(studentsTable.id, params.data.id));
  res.sendStatus(204);
});

router.get("/students/:id/gpa", async (req, res): Promise<void> => {
  const raw = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;
  const id = parseInt(raw, 10);
  if (isNaN(id)) {
    res.status(400).json({ error: "Invalid id" });
    return;
  }

  const [student] = await db.select().from(studentsTable).where(eq(studentsTable.id, id));
  if (!student) {
    res.status(404).json({ error: "Student not found" });
    return;
  }

  const enrollments = await db
    .select({ courseId: enrollmentsTable.courseId })
    .from(enrollmentsTable)
    .where(eq(enrollmentsTable.studentId, id));

  const courseGrades = [];

  for (const enr of enrollments) {
    const course = await db.select().from(coursesTable).where(eq(coursesTable.id, enr.courseId)).limit(1);
    if (!course[0]) continue;

    const assignments = await db.select().from(assignmentsTable).where(eq(assignmentsTable.courseId, enr.courseId));
    const scores: AssignmentScore[] = [];
    for (const asgn of assignments) {
      const gradeRows = await db.select().from(gradesTable).where(and(eq(gradesTable.studentId, id), eq(gradesTable.assignmentId, asgn.id))).limit(1);
      if (gradeRows[0]) {
        scores.push({
          assignmentId: asgn.id,
          score: Number(gradeRows[0].score),
          maxScore: Number(asgn.maxScore),
          weight: Number(asgn.weight),
          name: asgn.name,
          type: asgn.type,
        });
      }
    }

    const gradeInfo = StudentModel.computeCourseGrade(
      scores,
      course[0].id,
      course[0].name,
      course[0].code,
      course[0].credits,
      course[0].semester,
      course[0].gradingScheme as GradingScheme,
    );
    courseGrades.push(gradeInfo);
  }

  const dummy = new StudentModel(id, "", "", "", 0, "");
  const gpa = dummy.calculateGPA(courseGrades) ?? 0;
  const letter = scoreToLetter((gpa / 4) * 100);

  const breakdown = courseGrades
    .filter((c) => c.letterGrade !== null)
    .map((c) => ({
      course_name: c.courseName,
      grade: c.displayLabel ?? c.letterGrade,
      grade_points: c.gradePoints,
      credits: c.credits,
      grading_scheme: c.gradingScheme,
    }));

  res.json({
    student_id: id,
    gpa,
    total_courses: enrollments.length,
    completed_courses: breakdown.length,
    letter_grade: letter,
    grade_points_breakdown: breakdown,
  });
});

router.get("/students/:id/courses", async (req, res): Promise<void> => {
  const raw = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;
  const id = parseInt(raw, 10);
  if (isNaN(id)) {
    res.status(400).json({ error: "Invalid id" });
    return;
  }

  const enrollments = await db
    .select()
    .from(enrollmentsTable)
    .where(eq(enrollmentsTable.studentId, id));

  const result = [];
  for (const enr of enrollments) {
    const course = await db.select().from(coursesTable).where(eq(coursesTable.id, enr.courseId)).limit(1);
    if (!course[0]) continue;

    const assignments = await db.select().from(assignmentsTable).where(eq(assignmentsTable.courseId, enr.courseId));
    const scores: AssignmentScore[] = [];
    for (const asgn of assignments) {
      const gradeRows = await db.select().from(gradesTable).where(and(eq(gradesTable.studentId, id), eq(gradesTable.assignmentId, asgn.id))).limit(1);
      if (gradeRows[0]) {
        scores.push({
          assignmentId: asgn.id,
          score: Number(gradeRows[0].score),
          maxScore: Number(asgn.maxScore),
          weight: Number(asgn.weight),
          name: asgn.name,
          type: asgn.type,
        });
      }
    }

    const gradeInfo = StudentModel.computeCourseGrade(
      scores,
      course[0].id,
      course[0].name,
      course[0].code,
      course[0].credits,
      enr.semester,
      course[0].gradingScheme as GradingScheme,
    );

    result.push({
      course_id: gradeInfo.courseId,
      course_name: gradeInfo.courseName,
      course_code: gradeInfo.courseCode,
      semester: gradeInfo.semester,
      current_grade: gradeInfo.percentage,
      letter_grade: gradeInfo.letterGrade,
      display_label: gradeInfo.displayLabel,
      grading_scheme: gradeInfo.gradingScheme,
      credits: gradeInfo.credits,
    });
  }

  res.json(result);
});

export default router;

File: artifacts/api-server/src/routes/students.py
"""Student routes — full CRUD + GPA + courses."""
from typing import Optional, List
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from src.db.session import get_db
from src.db.models import StudentRow, EnrollmentRow, CourseRow, AssignmentRow, GradeRow
from src.domain.grade_utils import score_to_letter
from src.domain.student import Student, CourseGradeInfo
from src.domain.grade_book import AssignmentScore

router = APIRouter()


def _fmt_student(s: StudentRow, gpa=None):
    return {
        "id": s.id,
        "name": s.name,
        "email": s.email,
        "student_id": s.student_id,
        "year": s.year,
        "major": s.major,
        "clerk_user_id": s.clerk_user_id,
        "created_at": s.created_at.isoformat(),
        "gpa": gpa,
    }


def _compute_course_grade(db: Session, student_id: int, course: CourseRow, semester: str) -> CourseGradeInfo:
    assignments = db.query(AssignmentRow).filter(AssignmentRow.course_id == course.id).all()
    scores: List[AssignmentScore] = []
    for asgn in assignments:
        grade = db.query(GradeRow).filter(
            GradeRow.student_id == student_id,
            GradeRow.assignment_id == asgn.id,
        ).first()
        if grade:
            scores.append(AssignmentScore(
                assignment_id=asgn.id,
                score=float(grade.score),
                max_score=float(asgn.max_score),
                weight=float(asgn.weight),
                name=asgn.name,
                type=asgn.type,
            ))
    return Student.compute_course_grade(
        scores, course.id, course.name, course.code,
        course.credits, semester, course.grading_scheme or "weighted",
    )


def _compute_gpa(db: Session, student_id: int) -> Optional[float]:
    enrollments = db.query(EnrollmentRow).filter(EnrollmentRow.student_id == student_id).all()
    if not enrollments:
        return None
    course_grades = []
    for enr in enrollments:
        course = db.query(CourseRow).filter(CourseRow.id == enr.course_id).first()
        if not course:
            continue
        course_grades.append(_compute_course_grade(db, student_id, course, enr.semester))
    dummy = Student(student_id, "", "", "", 0, "")
    return dummy.calculate_gpa(course_grades)


class CreateStudentBody(BaseModel):
    name: str
    email: str
    student_id: str
    year: int
    major: str


class UpdateStudentBody(BaseModel):
    name: Optional[str] = None
    email: Optional[str] = None
    year: Optional[int] = None
    major: Optional[str] = None


@router.get("/students")
def list_students(
    search: Optional[str] = Query(None),
    major: Optional[str] = Query(None),
    year: Optional[int] = Query(None),
    db: Session = Depends(get_db),
):
    q = db.query(StudentRow)
    if search:
        q = q.filter(StudentRow.name.ilike(f"%{search}%"))
    if major:
        q = q.filter(StudentRow.major.ilike(f"%{major}%"))
    if year is not None:
        q = q.filter(StudentRow.year == year)
    students = q.order_by(StudentRow.name).all()
    return [_fmt_student(s, _compute_gpa(db, s.id)) for s in students]


@router.post("/students", status_code=201)
def create_student(body: CreateStudentBody, db: Session = Depends(get_db)):
    row = StudentRow(
        name=body.name, email=body.email, student_id=body.student_id,
        year=body.year, major=body.major,
    )
    db.add(row)
    db.commit()
    db.refresh(row)
    return _fmt_student(row, None)


@router.get("/students/{student_id}")
def get_student(student_id: int, db: Session = Depends(get_db)):
    s = db.query(StudentRow).filter(StudentRow.id == student_id).first()
    if not s:
        raise HTTPException(status_code=404, detail="Student not found")
    return _fmt_student(s, _compute_gpa(db, s.id))


@router.put("/students/{student_id}")
def update_student(student_id: int, body: UpdateStudentBody, db: Session = Depends(get_db)):
    s = db.query(StudentRow).filter(StudentRow.id == student_id).first()
    if not s:
        raise HTTPException(status_code=404, detail="Student not found")
    if body.name is not None:
        s.name = body.name
    if body.email is not None:
        s.email = body.email
    if body.year is not None:
        s.year = body.year
    if body.major is not None:
        s.major = body.major
    db.commit()
    db.refresh(s)
    return _fmt_student(s, _compute_gpa(db, s.id))


@router.delete("/students/{student_id}", status_code=204)
def delete_student(student_id: int, db: Session = Depends(get_db)):
    s = db.query(StudentRow).filter(StudentRow.id == student_id).first()
    if s:
        db.delete(s)
        db.commit()


@router.get("/students/{student_id}/gpa")
def get_student_gpa(student_id: int, db: Session = Depends(get_db)):
    s = db.query(StudentRow).filter(StudentRow.id == student_id).first()
    if not s:
        raise HTTPException(status_code=404, detail="Student not found")
    enrollments = db.query(EnrollmentRow).filter(EnrollmentRow.student_id == student_id).all()
    course_grades = []
    for enr in enrollments:
        course = db.query(CourseRow).filter(CourseRow.id == enr.course_id).first()
        if not course:
            continue
        course_grades.append(_compute_course_grade(db, student_id, course, enr.semester))
    dummy = Student(student_id, "", "", "", 0, "")
    gpa = dummy.calculate_gpa(course_grades) or 0.0
    letter = score_to_letter((gpa / 4) * 100)
    breakdown = [
        {
            "course_name": c.course_name,
            "grade": c.display_label or c.letter_grade,
            "grade_points": c.grade_points,
            "credits": c.credits,
            "grading_scheme": c.grading_scheme,
        }
        for c in course_grades if c.letter_grade is not None
    ]
    return {
        "student_id": student_id,
        "gpa": gpa,
        "total_courses": len(enrollments),
        "completed_courses": len(breakdown),
        "letter_grade": letter,
        "grade_points_breakdown": breakdown,
    }


@router.get("/students/{student_id}/courses")
def get_student_courses(student_id: int, db: Session = Depends(get_db)):
    enrollments = db.query(EnrollmentRow).filter(EnrollmentRow.student_id == student_id).all()
    result = []
    for enr in enrollments:
        course = db.query(CourseRow).filter(CourseRow.id == enr.course_id).first()
        if not course:
            continue
        gi = _compute_course_grade(db, student_id, course, enr.semester)
        result.append({
            "course_id": gi.course_id,
            "course_name": gi.course_name,
            "course_code": gi.course_code,
            "semester": gi.semester,
            "current_grade": gi.percentage,
            "letter_grade": gi.letter_grade,
            "display_label": gi.display_label,
            "grading_scheme": gi.grading_scheme,
            "credits": gi.credits,
        })
    return result
